### PR TITLE
refactor: migrated code to mvvm architecture

### LINKED ISSUE #231: Migrate Code to MVVM Architecture
**Describe the feature you'd like**
The Code should be migrated to model view viewmodel architecture

**Additional Context*
Here is the diagram depicting the same
picanim

**Would you like to work on it**
YES

### PR DESCRIPTION
**Pull request by adityastic**
Fixes #231 

Changes: 
- The Codebase is now migrated to MVVM Architecture
- MVVM got rid of the PreviewChangedListener
- Observables help us manage data change within our app without using listeners
- File Hierarchy is now according to the MVVM Architecture
- We use LiveData wherever possible

Additional Context:
Here is the diagram I tried to follow while migrating to this architecture
picanim

### CODE REVIEW

**Comment by adityastic** [comment]
@yashk2000 @anhanh11001 @ShridharGoel @liveHarshit @yatri1609 @naman653 @cweitat Please Review it and let me know if I missed anything or any thoughts on improving this PR.

**Comment by adityastic** [comment]
@liveHarshit great, I've mentioned them inside the message and also updated the pr name accordingly

File: app/src/main/java/com/nilhcem/blenamebadge/ui/custom/PreviewBadge.kt [hunk 1]
```diff
@@ -85,14 +88,14 @@ class PreviewBadge : View {
     override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
         super.onMeasure(widthMeasureSpec, heightMeasureSpec)
         val ratioHeight = 1
-        val ratioWidth = 3.42
```
Comment (yashk2000) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/custom/PreviewBadge.kt]: Was this required?

Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/custom/PreviewBadge.kt]: Yes, This is a must. 3.42 will have issues in larger xxxhdpi devices as the width is way too much. I can confirm this issue, as I own a xxxhdpi device

Comment (yashk2000) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/custom/PreviewBadge.kt]: Ok. I think the rest looks fine.

Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/custom/PreviewBadge.kt]: there will be a 14% decrease in height to be specific.

**Comment by yashk2000** [comment]
@iamareebjamal please review.

**Comment by iamareebjamal** [comment]
Very unrelated changes making it difficult to review

**Comment by liveHarshit** [comment]
@adityastic Can you please solve different issues in different PR. Also, you have not any issues about these changes.

**Comment by adityastic** [comment]
@iamareebjamal @liveHarshit sure, Let me separate everything other than MVVM from this PR

**Comment by adityastic** [comment]
@iamareebjamal I've updated the PR with the proper changes sir, also included a diagram which was used as a reference. @liveHarshit I talked about this in the comments of #219, Sorry for not opening a separate issue after the same. I've opened the specific issue and referenced it as well.
Please Review

Regards

File: app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/FragmentViewModelProviderFactory.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import com.nilhcem.blenamebadge.data.badge_preview.PreviewRepository
+import com.nilhcem.blenamebadge.data.fragments.FilesRepository
+import com.nilhcem.blenamebadge.ui.fragments.base.BaseFragmentViewModel
+
+class FragmentViewModelProviderFactory(private val configRepository: FilesRepository, private val previewRepo: PreviewRepository)
+    : ViewModelProvider.NewInstanceFactory() {
+
+    @Suppress("UNCHECKED_CAST")
```
Comment (liveHarshit) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/FragmentViewModelProviderFactory.kt]: Do not suppress unchecked cast.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
+package com.nilhcem.blenamebadge.ui.fragments.base
+
+import android.graphics.drawable.Drawable
```
Comment (liveHarshit) [inline app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt]: No android imports in the view model.

File: app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/FragmentViewModelProviderFactory.kt (hunk 2 above)
Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/FragmentViewModelProviderFactory.kt]: On it

File: app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt (hunk 3 above)
Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt]: Fixed them Sir, have a look

Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt]: I have a doubt though, what If I want to do processing on bitmaps or drawables and then assign it to a view. According to the pattern I should always work on my business logic in the DAO classes while it also does not allow you to pass the drawable or bitmap through the viewmodel, any suggestions would be appreciated sir :)

Comment (liveHarshit) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt]: You can do it in fragment itself, or can create a function in utils.

Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt]: @liveHarshit Alright, Let me improve the pr, give me 5 more minutes

**Comment by liveHarshit** [comment]
@iamareebjamal IMO, Codacy PR review should integrate for this project?

File: app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt (hunk 3 above)
Comment (adityastic) [reply app/src/main/java/com/nilhcem/blenamebadge/ui/fragments/base/BaseFragmentViewModel.kt]: @liveHarshit there was a potential context and uri leak in MainActivityViewHolder as well, Thanks for your advice, Ive improved the implementation. Please have a look

**Comment by adityastic** [comment]
@liveHarshit that would be great, will help us to improve our code quality

**Comment by liveHarshit** [comment]
@iamareebjamal What is the codecy logic for showing undocumented public classes and functions?

**Comment by liveHarshit** [comment]
@adityastic Do not use unsafe casting, you can use kotlin smart cast instead or avoid casting by changing in logic.

**Comment by liveHarshit** [comment]
> @adityastic Do not use unsafe casting, you can use kotlin smart cast instead or avoid casting by changing in the logic.

**Comment by liveHarshit** [comment]
> @iamareebjamal What is the reason for showing undocumented public classes and functions in codacy review?

**Comment by adityastic** [comment]
@liveHarshit Done, fixed them. Have a look

**Comment by adityastic** [comment]
@iamareebjamal please review sir

**Comment by adityastic** [comment]
@iamareebjamal @liveHarshit Codacy/PR specified the suggestions errors:
1. TooManyFunctions - Interfaces and Activities need to have those functions, If I even try to split the functions to two interfaces it will still end up in the activity which also has this suggestions
2. LongParameterList - Because the no of parameters is more than 6, Even If I create a data class with the same objects it will show these errors in those data classes.

What should I do about the Codacy/PR Quality Review?

**Comment by liveHarshit** [comment]
> What should I do about the Codacy/PR Quality Review?

Do not cast. And for long parameters use a new line. Reset good. :+1:

**Comment by adityastic** [comment]
@liveHarshit on it. Just a min

**Comment by iamareebjamal** [comment]
> Even If I create a data class with the same objects it will show these errors in those data classes.

No , it won't

**Comment by adityastic** [comment]
@liveHarshit Build was Successful, Please Review. @iamareebjamal you are amazing sir, it worked like a charm, thanks for the suggestion, Ive updated the PR please review

**Comment by adityastic** [comment]
@iamareebjamal please review sir, peer review done

**Pull request merged**